core/src/main/scala/cats/data/Catenable.scala

```diff
@@ -0,0 +1,473 @@
+ * O(1) `uncons`, such that walking the sequence via N successive `uncons`
+ * steps takes O(N). Like a difference list, conversion to a `Seq[A]`
+ * takes linear time, regardless of how the sequence is built up.
+ */
+sealed abstract class Catenable[+A] {
+
+  /** Returns the head and tail of this catenable if non empty, none otherwise. Amortized O(1). */
+  final def uncons: Option[(A, Catenable[A])] = {
+    var c: Catenable[A] = this
+    val rights = new collection.mutable.ArrayBuffer[Catenable[A]]
+    // scalastyle:off null
+    var result: Option[(A, Catenable[A])] = null
+    while (result eq null) {
+      c match {
+        case Singleton(a) =>
+          val next =
+            if (rights.isEmpty) nil
+            else rights.reduceLeft((x, y) => Append(y, x))
+          result = Some(a -> next)
+        case Append(l, r) => c = l; rights += r
+        case Wrap(seq) =>
+          val tail = seq.tail
+          val next = fromSeq(tail)
+          result = Some(seq.head -> next)
+        case Empty =>
+          if (rights.isEmpty) {
+            result = None
+          } else {
+            c = rights.last
+            rights.trimEnd(1)
+          }
+      }
+    }
+    // scalastyle:on null
+    result
+  }
+
+  /** Returns true if there are no elements in this collection. */
+  def isEmpty: Boolean
+
+  /** Returns false if there are no elements in this collection. */
+  def nonEmpty: Boolean = !isEmpty
+
+  /** Concatenates this with `c` in O(1) runtime. */
+  final def ++[A2 >: A](c: Catenable[A2]): Catenable[A2] =
+    append(this, c)
+
+  /** Returns a new catenable consisting of `a` followed by this. O(1) runtime. */
+  final def cons[A2 >: A](a: A2): Catenable[A2] =
+    append(singleton(a), this)
+
+  /** Alias for [[cons]]. */
+  final def +:[A2 >: A](a: A2): Catenable[A2] =
+    cons(a)
+
+  /** Returns a new catenable consisting of this followed by `a`. O(1) runtime. */
+  final def snoc[A2 >: A](a: A2): Catenable[A2] =
+    append(this, singleton(a))
+
+  /** Alias for [[snoc]]. */
+  final def :+[A2 >: A](a: A2): Catenable[A2] =
+    snoc(a)
+
+  /** Applies the supplied function to each element and returns a new catenable. */
+  final def map[B](f: A => B): Catenable[B] =
+    fromSeq(iterator.map(f).toVector)
+
+  /** Applies the supplied function to each element and returns a new catenable from the concatenated results */
+  final def flatMap[B](f: A => Catenable[B]): Catenable[B] =
+    foldLeft(nil: Catenable[B])((acc, a) => acc ++ f(a))
+
+  /** Folds over the elements from left to right using the supplied initial value and function. */
+  final def foldLeft[B](z: B)(f: (B, A) => B): B = {
+    var result = z
+    foreach(a => result = f(result, a))
+    result
+  }
+
+  /** Folds over the elements from right to left using the supplied initial value and function. */
+  final def foldRight[B](z: B)(f: (A, B) => B): B = {
+    val stack = new ArrayStack[A]
+    foreach { a => stack += a; () }
```

Comment:
Can we use iterator here to avoid the closure. Note chain had a reverseIterator which means we don’t need to allocate a duplicate.